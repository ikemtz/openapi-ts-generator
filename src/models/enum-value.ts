export interface IEnumValue {
  key?: number;
  name: string;
  titleName: string;
  snakeCaseName: string;
}
