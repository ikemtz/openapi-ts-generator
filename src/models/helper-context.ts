import { IEntity } from './template-data';

export interface IHelperContext {
  data: {
    root: IEntity;
    key: number;
    index: number;
    first: boolean;
  };
  name: string;
}

export type PropertyType = 'valueProperties' | 'referenceProperties';
