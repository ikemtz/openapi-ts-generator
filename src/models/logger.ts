export interface ILogger {
  log: (data: string) => void;
  error: (data: string) => void;
  warn: (data: string) => void;
}

export class MockConsoleLogger implements ILogger {
  // tslint:disable-next-line: no-empty
  public log = (data: string) => {};
  // tslint:disable-next-line: no-empty
  public error = (data: string) => {};
  // tslint:disable-next-line: no-empty
  public warn = (data: string) => {};
}
