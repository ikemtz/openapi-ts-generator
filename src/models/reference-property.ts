export interface IReferenceProperty {
  name: string;
  snakeCaseName: string;
  referenceTypeName: string;
  typeScriptType: string;
  hasValidators: boolean;
  isSameAsParentTypescriptType: boolean;
  hasMultipleValidators: boolean;
  maxItems?: number;
  minItems?: number;
  maxLength?: number;
  minLength?: number;
  maximum?: number;
  minimum?: number;
  isArray: boolean;
  required: boolean;
  isEnum?: boolean;
  initialValue: string;
  initialTestValue: string;
}
