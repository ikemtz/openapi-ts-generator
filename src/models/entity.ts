import { IReferenceProperty } from './reference-property';
import { IValueProperty } from './value-property';
import { IImportType } from './template-data';

export interface IEntity {
  isEnum?: boolean;
  enumValues: (string | { key?: number; name: string; titleName: string })[];
  name: string;
  kebabCasedName: string;
  camelSingularName: string;
  description?: string;
  importTypes: IImportType[];
  valueProperties: IValueProperty[];
  referenceProperties: IReferenceProperty[];
}
