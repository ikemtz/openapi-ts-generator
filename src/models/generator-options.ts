/* eslint-disable @typescript-eslint/no-unused-vars */
import { resolve } from 'path';
import { ILogger } from './logger';
import { IEntity } from './entity';
import { IReferenceProperty } from './reference-property';
import { IValueProperty } from './value-property';

export interface IGeneratorOptions {
  logger?: ILogger;
  outputPath: string;
  openApiJsonUrl?: string;
  openApiJsonFileName?: string;
  genAngularFormGroups?: boolean;
  typeFilterCallBack?: (entity: IEntity, index: number, array: IEntity[]) => boolean;
  valuePropertyTypeFilterCallBack?: (valueProperty: IValueProperty, index: number, array: IValueProperty[]) => boolean;
  referencePropertyTypeFilterCallBack?: (referenceProperty: IReferenceProperty, index: number, array: IReferenceProperty[]) => boolean;
  pathUrlFormattingCallBack?: (url: string) => string;
  templates?: ITemplates | null;
}
export interface ITemplates {
  model?: string;
  testObjectFactory?: string;
  form?: string;
  formGroupFactory?: string;
  modelProperties?: string;
  barrel?: string;
  enum?: string;
  endpoints?: string;
}
export function defaultFilter(
  _value: IEntity | IValueProperty | IReferenceProperty,
  _index: number,
  _array: IEntity[] | IValueProperty[] | IReferenceProperty[],
): boolean {
  return true;
}

export function setGeneratorOptionDefaults(options: IGeneratorOptions): IGeneratorOptions {
  const templateFolder = resolve(`${__dirname}/..`, 'templates');
  options.typeFilterCallBack = options.typeFilterCallBack ?? defaultFilter;
  options.valuePropertyTypeFilterCallBack = options.valuePropertyTypeFilterCallBack ?? defaultFilter;
  options.referencePropertyTypeFilterCallBack = options.referencePropertyTypeFilterCallBack ?? defaultFilter;
  options.templates = {
    ...options.templates,
    model: options.templates?.model ?? `${templateFolder}/model.ts.hbs`,
    form: options.templates?.form ?? `${templateFolder}/form.ts.hbs`,
    formGroupFactory: options.templates?.formGroupFactory ?? `${templateFolder}/form-group-factory.ts.hbs`,
    testObjectFactory: options.templates?.testObjectFactory ?? `${templateFolder}/test-object-factory.ts.hbs`,
    barrel: options.templates?.barrel ?? `${templateFolder}/index.ts.hbs`,
    enum: options.templates?.barrel ?? `${templateFolder}/enum.ts.hbs`,
    modelProperties: options.templates?.barrel ?? `${templateFolder}/model-properties.ts.hbs`,
    endpoints: options.templates?.endpoints ?? `${templateFolder}/endpoints.ts.hbs`,
  };
  return options;
}
