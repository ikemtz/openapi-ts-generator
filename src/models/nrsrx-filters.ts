/* eslint-disable @typescript-eslint/no-unused-vars */
import { IEntity } from './entity';
import { IValueProperty } from './value-property';
const edmRegex = /^Edm[A-z]*Kind$/;

export function nrsrxTypeFilterCallBack(val: IEntity, _i: number, _arr: IEntity[]): boolean {
  return !val.name.endsWith('ODataEnvelope') && !edmRegex.test(val.name);
}
export function nrsrxValuePropertyTypeFilterCallBack(val: IValueProperty, _i: number, _arr: IValueProperty[]): boolean {
  return !val.name.startsWith('created') && !val.name.startsWith('updated');
}
