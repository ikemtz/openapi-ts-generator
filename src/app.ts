import { generateTsModels } from '.';
import { IGeneratorOptions } from './models/generator-options';
import { nrsrxTypeFilterCallBack, nrsrxValuePropertyTypeFilterCallBack } from './models/nrsrx-filters';
import { IEntity } from './models/template-data';

export const unitGenerationOptions = {
  openApiJsonUrl: 'https://d-unit-wal-00-cus-mstrcrp.azurewebsites.net/swagger/v1/swagger.json',
  outputPath: './output_unit/',
  typeFilterCallBack: nrsrxTypeFilterCallBack,
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
};
export const messageGenerationOptions = {
  openApiJsonUrl: 'https://d-msng-wal-01-cus-mstrcrp.azurewebsites.net/swagger/v1/swagger.json',
  outputPath: './output_msng/',
  typeFilterCallBack: (val: IEntity, i: number, arr: IEntity[]) =>
    nrsrxTypeFilterCallBack(val, i, arr) && val.name !== 'GetMessageInfoResponse',
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
};
export const accountGenerationOptions = {
  openApiJsonUrl: 'https://d-acct-wal-01-cus-mstrcrp.azurewebsites.net/swagger/v1/swagger.json',
  outputPath: './output_acct/',
  typeFilterCallBack: nrsrxTypeFilterCallBack,
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
};

export const fileEmployeeGenerationOptions: IGeneratorOptions = {
  openApiJsonFileName: '../src/test-swagger-document.json',
  outputPath: './output_empl/',
  typeFilterCallBack: nrsrxTypeFilterCallBack,
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
};

export function generateUnitFiles(options: IGeneratorOptions = unitGenerationOptions) {
  return generateTsModels(options);
}

export function generateMessageFiles(options: IGeneratorOptions = messageGenerationOptions) {
  return generateTsModels(messageGenerationOptions);
}

export function generateAccountFiles(options: IGeneratorOptions = accountGenerationOptions) {
  return generateTsModels(accountGenerationOptions);
}

export function generateEmployeeFiles(options: IGeneratorOptions = fileEmployeeGenerationOptions) {
  return generateTsModels(fileEmployeeGenerationOptions);
}

generateUnitFiles();
generateMessageFiles();
generateAccountFiles();
generateEmployeeFiles();
