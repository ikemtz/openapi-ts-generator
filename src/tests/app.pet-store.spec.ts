import { mkdirSync } from 'fs';
import { generateTsModels, nrsrxTypeFilterCallBack, nrsrxValuePropertyTypeFilterCallBack } from '..';
import { ValidateFiles } from './app.spec';
import { IGeneratorOptions } from '../models/generator-options';

const accountGenerationOptionsFactory = (): IGeneratorOptions => ({
  openApiJsonUrl: 'https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.json',
  outputPath: './jest_output/pet-store/',
  typeFilterCallBack: nrsrxTypeFilterCallBack,
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
  genAngularFormGroups: true,
});

describe('Url Based - Full Integration Tests', () => {
  describe('Sample PetStore Service', () => {
    it('should generate files', async () => {
      const options = accountGenerationOptionsFactory();
      try {
        mkdirSync(options.outputPath, { recursive: true });
      } catch {
        // ignore
      }
      await generateTsModels(options);
      ValidateFiles(options);
    });
  });
});
