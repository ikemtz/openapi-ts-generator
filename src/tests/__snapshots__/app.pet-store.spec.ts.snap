// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files 1`] = `
Array [
  "endpoints.ts",
  "error.form-group-fac.ts",
  "error.model.ts",
  "error.properties.ts",
  "index.ts",
  "pet.form-group-fac.ts",
  "pet.model.ts",
  "pet.properties.ts",
  "pets.model.ts",
]
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: endpoints.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #endpoints-ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
/* tslint:disable */
export enum Endpoints {
  pets = '/pets',
  pets2 = '/pets/{petId}',
}
export interface IEndpoints {
  readonly pets: string;
  readonly pets2: string;
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: error.form-group-fac.ts 1`] = `
"/* eslint-disable @typescript-eslint/no-unused-vars */
/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #form-group-factory.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
import { FormControl, FormArray, FormGroup, Validators } from '@angular/forms'; //NOSONAR

export function ErrorFormGroupFac(): FormGroup {
  return new FormGroup({
    code: new FormControl<number>(0, { validators: Validators.required, nonNullable: true } ),
    message: new FormControl<string>('', { validators: Validators.required, nonNullable: true } ),
  });
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: error.model.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #model.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IError {
  code?: number;
  message?: string;
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: error.properties.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #model-properties.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
/* tslint:disable */
export enum ErrorProperties {
  CODE = 'code',
  MESSAGE = 'message',
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: index.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #generate-barrel-ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
/* tslint:disable */
export * from './error.form-group-fac';
export * from './error.model';
export * from './error.properties';
export * from './pet.form-group-fac';
export * from './pet.model';
export * from './pet.properties';
export * from './pets.model';
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: pet.form-group-fac.ts 1`] = `
"/* eslint-disable @typescript-eslint/no-unused-vars */
/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #form-group-factory.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
import { FormControl, FormArray, FormGroup, Validators } from '@angular/forms'; //NOSONAR

export function PetFormGroupFac(): FormGroup {
  return new FormGroup({
    id: new FormControl<number | null>(null),
    name: new FormControl<string>('', { validators: Validators.required, nonNullable: true } ),
    tag: new FormControl<string | null>(null),
  });
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: pet.model.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #model.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IPet {
  id?: number;
  name?: string;
  tag?: string;
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: pet.properties.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #model-properties.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
/* tslint:disable */
export enum PetProperties {
  ID = 'id',
  NAME = 'name',
  TAG = 'tag',
}
"
`;

exports[`Url Based - Full Integration Tests Sample PetStore Service should generate files: pets.model.ts 1`] = `
"/* istanbul ignore file */
/**
 * This file is generated by the openapi-ts-generator
 * #model.ts.hbs
 * For issues or feature request, visit the repo: https://github.com/ikemtz/openapi-ts-generator
 * Do not edit.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IPets {
}
"
`;
